### PR TITLE
Fix handling of unrecognized naming style values

### LINKED ISSUE #18121: Cannot create naming rules for type parameters and local variables
In trying to configure naming rules in an .editorconfig, I've found that the current implementation of naming rules lacks the ability to target certain symbol kinds. In my scenario I need the ability to target type parameters and local variables.

Type parameters and variables should be supported symbol kinds both through the options UI in VS and through .editorconfig files.

### LINKED ISSUE #20907: Editorconfig backwards compatibility dilemma
We have a bit of a dilemma. If you don't fix this in 15.0–15.2, it will still get in the way and totally break the experience for anyone using 15.0–15.2 if we ever use `type_parameter` in an OSS project's .editorconfig.

Imagine that 15.3 is out with support for `type_parameter` and we want to enforce a `T` prefix followed by pascal case:
```ini
dotnet_naming_style.type_parameter.required_prefix = T
dotnet_naming_style.type_parameter.capitalization = pascal_case 
dotnet_naming_symbols.type_parameters.applicable_kinds = type_parameter
dotnet_naming_rule.type_parameters.severity = error
dotnet_naming_rule.type_parameters.symbols = type_parameters
dotnet_naming_rule.type_parameters.style = type_parameter
```

The problem is that this same block of settings in VS15.0–15.2 acts as though you had said `applicable_kinds = *`. This is a problem because suddenly every identifier ever is required by VS15.0–15.2 to begin with `T`.

This is a very similar problem to[ https://github.com/dotnet/roslyn/issues/20891](dotnet_naming_rule.*.symbols).

The only fix I can see is to push a fix to VS15.0–15.2 that
 1. splits the line by commas
 2. **removes the items it doesn't recognize**
 3. **If the line exists and the list of recognized items is empty, apply the rule to nothing instead of applying the rule to everything.**

For `applicable_kinds`, you have 1 and 2 correct but you can verify that you don't have 3 correct because `type_parameter, event` acts as `event` but `type_parameter` acts as `*`.

For `dotnet_naming_rule.*.symbols`, as far as I can tell, you have all three parts wrong for backwards compatibility.

Otherwise, as you version this very useful and attractive system, how is anyone supposed to adopt new features without forcing everyone in the codebase to upgrade to the very latest VS before continuing any work, or bear through a flood of erroneous squiggles?

Thanks for considering.

### LINKED ISSUE #23336: EditorConfig Naming : type_parameter is typed to namespace?
Was working on building out some naming conventions and realized that something fishy is going on with the symbols `type_parameter` and `namespace`. `namespace` doesn't seem to be a recognized symbol and rules defined for `type_parameters` run on namespace symbols (rather than parameters).

```
# Style Definitions
dotnet_naming_style.pascal_case_style.capitalization            = pascal_case
dotnet_naming_style.camel_case_style.capitalization             = camel_case

# Classes, structs, methods, enums, events, properties, namespaces, delegates must be PascalCase
dotnet_naming_rule.general_naming.severity                  = suggestion
dotnet_naming_rule.general_naming.symbols                   = general
dotnet_naming_rule.general_naming.style                     = pascal_case_style
dotnet_naming_symbols.general.applicable_kinds              = class, namespace
dotnet_naming_symbols.general.applicable_accessibilities    = *

# Everything else is camelCase
dotnet_naming_rule.everything_else_naming.severity                  = suggestion
dotnet_naming_rule.everything_else_naming.symbols                   = everything_else
dotnet_naming_rule.everything_else_naming.style                     = camel_case_style
dotnet_naming_symbols.everything_else.applicable_kinds              = type_parameter
dotnet_naming_symbols.everything_else.applicable_accessibilities       = *
```

### LINKED ISSUE #23709: Cannot refer to "private protected" in "applicable_accessibilities"
```
dotnet_naming_rule.externally_visible_members_must_be_pascal_cased.symbols                                      = externally_visible_symbols
dotnet_naming_rule.externally_visible_members_must_be_pascal_cased.style                                        = pascal_case_style
dotnet_naming_rule.externally_visible_members_must_be_pascal_cased.severity                                     = warning

# Symbols
dotnet_naming_symbols.externally_visible_symbols.applicable_kinds                                               = field
dotnet_naming_symbols.externally_visible_symbols.applicable_accessibilities                                     = private_protected

# Styles
dotnet_naming_style.pascal_case_style.capitalization                                                            = pascal_case
```

``` C#
    class Program
    {
        private protected string fooBar;
    }
```

Expected: Warning on the naming
Actual: No warning

### LINKED ISSUE #24248: .editorconfig some values aren't recognizable/supported by VS
**Version Used**: Visual Studio Community 2017 v15.5.3

1. `flush_left` is documented [here](https://github.com/MicrosoftDocs/visualstudio-docs/blob/master/docs/ide/editorconfig-code-style-settings-reference.md) but [here](https://gist.github.com/jmarolf/0e165b27a498b49e3b3512004ea9d8b8) it's documented as `left_most` and yet VS doesn't recognize either of these values.

2. `type_parameter` and `namespace` they are documented [here](https://gist.github.com/jmarolf/0e165b27a498b49e3b3512004ea9d8b8) but are not documented anywhere else. @jmarolf does VS really support these values?

3. <s>`when_on_single_line` is documented [here](https://github.com/MicrosoftDocs/visualstudio-docs/blob/master/docs/ide/editorconfig-code-style-settings-reference.md) but it's not mentioned [here](https://gist.github.com/jmarolf/0e165b27a498b49e3b3512004ea9d8b8) and VS doesn't recognize it.</s> Resolved.

It can be really confusing because you don't know whether VS supports these values or they are just aren't recognizable by the IDE and these values are important to my style of coding, probably to others too.

### PR DESCRIPTION
**Pull request by sharwell**
* The first commit implements https://github.com/dotnet/roslyn/issues/20907#issuecomment-385058210
* The second commit fixes #18121, fixes #23336, fixes #23709, and closes #24248
* The third commit implements https://github.com/dotnet/roslyn/issues/20907#issuecomment-387071711, which closes #20907

### Baseline (15.8 Preview 2)

image

image

### Baseline (current master branch e3dcf8c76dc2c228f61b9de8af92f1ec0275f5f8)

image

image

### After second commit

image

image

### After third commit

image

image

### Customer scenario

A user attempts to create a naming rule in .editorconfig for a symbol type added in a recent release of Roslyn. Old versions treat the rule as apply to *all* symbols when it should apply to *no* symbols. A similar situation applies for accessibilities.

In addition, some of our documented member kinds (namespace and type_parameter) do not work as documented.

### Bugs this fixes

#18121
#20907
#23336
#23709
#24248

### Workarounds, if any

Don't use the new feature.

### Risk

This change is relatively low risk for Preview 3. It builds on other new feature work in Preview 3, and helps ensure that a consistent experience is provided for the new feature now and in future releases.

### Performance impact

Low. Existing working patterns were used for the new analysis.

### Is this a regression from a previous update?

No.

### Root cause analysis

This is a supporting change for feature work already in Preview 3. It was intended for inclusion in the "mop-up" on Monday, but infrastructure and test problems delayed the work. Now that we see the tests working again with this pull request we are ready to submit it.

### How was the bug found?

Customer and internal customer reported.

### Test documentation updated?

No, but many new tests are added.

### CODE REVIEW

File: src/Workspaces/Core/Portable/NamingStyles/Serialization/AccessibilityExtensions.cs [hunk 1]
```diff
@@ -21,5 +21,30 @@ internal static Accessibility FromXElement(XElement accessibilityElement)
...
+                switch (currentSymbol.Kind)
+                {
+                case SymbolKind.Namespace:
+                    return Accessibility.Public;
+
+                case SymbolKind.Parameter:
+                case SymbolKind.TypeParameter:
+                    continue;
+
+                case SymbolKind.Method when ((IMethodSymbol)currentSymbol).MethodKind == MethodKind.LocalFunction:
```
Comment (Neme12) [inline src/Workspaces/Core/Portable/NamingStyles/Serialization/AccessibilityExtensions.cs]: What about parameters of lambdas & anonymous methods?
They should be local, so this is likely missing another case for `MethodKind`.

Comment (sharwell) [reply src/Workspaces/Core/Portable/NamingStyles/Serialization/AccessibilityExtensions.cs]: I added the following tests to `TestPascalCaseSymbol_ExpectedSymbolAndAccessibility` and fixed the condition here.

```
[InlineData("void Outer() { void Inner(int [|m|]) {} }", "void Outer() { void Inner(int M) {} }", SymbolKind.Parameter, Accessibility.NotApplicable)]
[InlineData("void Outer() { System.Action<int> action = [|m|) => {} }", "void Outer() { System.Action<int> action = M => {} }", SymbolKind.Parameter, Accessibility.NotApplicable)]
[InlineData("void Outer() { System.Action<int> action = ([|m|]) => {} }", "void Outer() { System.Action<int> action = (M) => {} }", SymbolKind.Parameter, Accessibility.NotApplicable)]
[InlineData("void Outer() { System.Action<int> action = (int [|m|]) => {} }", "void Outer() { System.Action<int> action = (int M) => {} }", SymbolKind.Parameter, Accessibility.NotApplicable)]
[InlineData("void Outer() { System.Action<int> action = delegate (int [|m|]) {} }", "void Outer() { System.Action<int> action = delegate (int M) {} }", SymbolKind.Parameter, Accessibility.NotApplicable)]
```

I'll wait to see if there are other tests I need to add before pushing this update.

Comment (sharwell) [reply src/Workspaces/Core/Portable/NamingStyles/Serialization/AccessibilityExtensions.cs]: ➡️ Fixed in 40ded3c5678be13246ac7a017864a24cf0a890f5

File: src/EditorFeatures/CSharpTest/Diagnostics/NamingStyles/NamingStylesTests.cs [hunk 2]
```diff
@@ -55,6 +57,128 @@ internal override (DiagnosticAnalyzer, CodeFixProvider) CreateDiagnosticProvider
...
+        [InlineData("void Outer() { System.Action<int> action = ([|m|]) => {} }", "void Outer() { System.Action<int> action = (M) => {} }", SymbolKind.Parameter, Accessibility.NotApplicable)]
+        [InlineData("void Outer() { System.Action<int> action = (int [|m|]) => {} }", "void Outer() { System.Action<int> action = (int M) => {} }", SymbolKind.Parameter, Accessibility.NotApplicable)]
+        [InlineData("void Outer() { System.Action<int> action = delegate (int [|m|]) {} }", "void Outer() { System.Action<int> action = delegate (int M) {} }", SymbolKind.Parameter, Accessibility.NotApplicable)]
+        [InlineData("void Outer() { int [|m|]; }", "void Outer() { int M; }", SymbolKind.Local, Accessibility.NotApplicable)]
+        [WorkItem(20907, "https://github.com/dotnet/roslyn/issues/20907")]
+        public async Task TestPascalCaseSymbol_ExpectedSymbolAndAccessibility(string camelCaseSymbol, string pascalCaseSymbol, object symbolKind, Accessibility accessibility)
+        {
+            var alternateSymbolKind = TypeKind.Class.Equals(symbolKind) ? TypeKind.Interface : TypeKind.Class;
```
Comment (heejaechang) [inline src/EditorFeatures/CSharpTest/Diagnostics/NamingStyles/NamingStylesTests.cs]: I am fine with this InlineData. but with this many inlineData, how one debug it? should one go through every inlinedata until right inline data is finally used for the test?

seems it is just different way to do stuff than our usual test just calling VerifyXXXX(code, epected result);

Comment (sharwell) [reply src/EditorFeatures/CSharpTest/Diagnostics/NamingStyles/NamingStylesTests.cs]: > with this many inlineData, how one debug it?

You can right click the specific case in Test Explorer and **Debug Test**.

File: src/EditorFeatures/CSharpTest/Diagnostics/NamingStyles/EditorConfigNamingStyleParserTests.cs [hunk 3]
```diff
@@ -12,6 +13,24 @@ namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.Diagnostics.NamingStyle
 {
     public class EditorConfigNamingStyleParserTests
     {
+        internal static SymbolKindOrTypeKind ToSymbolKindOrTypeKind(object symbolOrTypeKind)
+        {
+            switch (symbolOrTypeKind)
+            {
+            case TypeKind typeKind:
```
Comment (dpoeschl) [inline src/EditorFeatures/CSharpTest/Diagnostics/NamingStyles/EditorConfigNamingStyleParserTests.cs]: Looks like your indentation settings might be off (I think it's usually one deeper)

File: src/EditorFeatures/CSharpTest/Diagnostics/NamingStyles/EditorConfigNamingStyleParserTests.cs [hunk 4]
```diff
@@ -298,31 +325,64 @@ public static void TestNoRulesAreReturned()
...
-        public static void TestApplicableAccessibilitiesParse()
+        [Theory]
+        [InlineData("property,method", new object[] { SymbolKind.Property, MethodKind.Ordinary })]
+        [InlineData("namespace", new object[] { SymbolKind.Namespace })]
+        [InlineData("type_parameter", new object[] { SymbolKind.TypeParameter })]
+        [InlineData("*", new object[] { SymbolKind.Namespace, TypeKind.Class, TypeKind.Struct, TypeKind.Interface, TypeKind.Enum, SymbolKind.Property, MethodKind.Ordinary, MethodKind.LocalFunction, SymbolKind.Field, SymbolKind.Event, TypeKind.Delegate, SymbolKind.Parameter, SymbolKind.TypeParameter, SymbolKind.Local })]
+        [InlineData(null, new object[] { SymbolKind.Namespace, TypeKind.Class, TypeKind.Struct, TypeKind.Interface, TypeKind.Enum, SymbolKind.Property, MethodKind.Ordinary, MethodKind.LocalFunction, SymbolKind.Field, SymbolKind.Event, TypeKind.Delegate, SymbolKind.Parameter, SymbolKind.TypeParameter, SymbolKind.Local })]
+        [InlineData("property,method,invalid", new object[] { SymbolKind.Property, MethodKind.Ordinary })]
+        [InlineData("invalid", new object[] { })]
+        [InlineData("", new object[] { })]
```
Comment (DustinCampbell) [inline src/EditorFeatures/CSharpTest/Diagnostics/NamingStyles/EditorConfigNamingStyleParserTests.cs]: nice!

File: src/EditorFeatures/CSharpTest/Diagnostics/NamingStyles/EditorConfigNamingStyleParserTests.cs [hunk 5]
```diff
@@ -298,31 +325,64 @@ public static void TestNoRulesAreReturned()
...
+        [Theory]
+        [InlineData("internal,protected_internal", new[] { Accessibility.Internal, Accessibility.ProtectedOrInternal })]
+        [InlineData("friend,protected_friend", new[] { Accessibility.Friend, Accessibility.ProtectedOrFriend })]
+        [InlineData("private_protected", new[] { Accessibility.ProtectedAndInternal })]
+        [InlineData("local", new[] { Accessibility.NotApplicable })]
+        [InlineData("*", new[] { Accessibility.NotApplicable, Accessibility.Public, Accessibility.Internal, Accessibility.Private, Accessibility.Protected, Accessibility.ProtectedAndInternal, Accessibility.ProtectedOrInternal })]
+        [InlineData(null, new[] { Accessibility.NotApplicable, Accessibility.Public, Accessibility.Internal, Accessibility.Private, Accessibility.Protected, Accessibility.ProtectedAndInternal, Accessibility.ProtectedOrInternal })]
+        [InlineData("internal,protected,invalid", new[] { Accessibility.Internal, Accessibility.Protected })]
+        [InlineData("invalid", new Accessibility[] { })]
+        [InlineData("", new Accessibility[] { })]
```
Comment (DustinCampbell) [inline src/EditorFeatures/CSharpTest/Diagnostics/NamingStyles/EditorConfigNamingStyleParserTests.cs]: nice again!

File: src/EditorFeatures/CSharpTest/Diagnostics/NamingStyles/EditorConfigNamingStyleParserTests.cs (hunk 3 above)
Comment (sharwell) [reply src/EditorFeatures/CSharpTest/Diagnostics/NamingStyles/EditorConfigNamingStyleParserTests.cs]: Same comment as https://github.com/dotnet/roslyn/pull/26589#discussion_r193532991

Comment (sharwell) [reply src/EditorFeatures/CSharpTest/Diagnostics/NamingStyles/EditorConfigNamingStyleParserTests.cs]: I'll update as a separate PR so we don't have to retest this one for just indentation.

Comment (sharwell) [reply src/EditorFeatures/CSharpTest/Diagnostics/NamingStyles/EditorConfigNamingStyleParserTests.cs]: :arrow_right: Had to rerun tests for retargeting, so fixed in 32e150c9ae85d9ad3e3233ea374e3ae0cab26d3b

**Comment by sharwell** [comment]
@dotnet-bot retest this please

**Comment by sharwell** [comment]
@dotnet-bot retest this please

**Comment by jinujoseph** [comment]
Approved to merge for 15.8.Preview4 (Once tests are green)

**Comment by sharwell** [comment]
Both integration test failures are the issue from #27933

**Pull request merged**